src/mlpack/methods/reinforcement_learning/q_networks/categorical_dqn.hpp

```diff
@@ -0,0 +1,224 @@
+/**
+ * @file methods/reinforcement_learning/q_networks/categorical_dqn.hpp
+ * @author Nishant Kumar
+ *
+ * This file contains the implementation of the categorical deep q network.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_CATEGORICAL_DQN_HPP
+#define MLPACK_METHODS_RL_CATEGORICAL_DQN_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/init_rules/gaussian_init.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/loss_functions/mean_squared_error.hpp>
+#include <mlpack/methods/ann/loss_functions/empty_loss.hpp>
+
+namespace mlpack {
+namespace rl {
+
+using namespace mlpack::ann;
+
+/**
+ * @tparam OutputLayerType The output layer type of the network.
+ * @tparam InitType The initialization type used for the network.
+ * @tparam NetworkType The type of network used for simple dqn.
+ */
+template<
+  typename OutputLayerType = EmptyLoss<>,
+  typename InitType = GaussianInitialization,
+  typename NetworkType = FFN<OutputLayerType, InitType>
+>
+class CategoricalDQN
+{
+ public:
+  /**
+   * Default constructor.
+   */
+  CategoricalDQN() : network(), isNoisy(false), atomSize(0)
+  { /* Nothing to do here. */ }
+
+  /**
+   * Construct an instance of CategoricalDQN class.
+   *
+   * @param inputDim Number of inputs.
+   * @param h1 Number of neurons in hiddenlayer-1.
+   * @param h2 Number of neurons in hiddenlayer-2.
+   * @param outputDim Number of neurons in output layer.
+   * @param isNoisy Specifies whether the network needs to be of type noisy.
+   * @param init Specifies the initialization rule for the network.
+   * @param outputLayer Specifies the output layer type for network.
+   * @param atomSize Specifies the number of atoms to be used.
+   */
+  CategoricalDQN(const int inputDim,
+            const int h1,
+            const int h2,
+            const int outputDim,
+            const bool isNoisy = false,
+            InitType init = InitType(),
+            OutputLayerType outputLayer = OutputLayerType(),
+            const size_t atomSize = 51):
```

Comment:
`atomSize` is a parameter, but it is hardcoded in the `TrainCategoricalAgent()` function. Am I misunderstanding something, or should we change that?

Reply:
Oh sorry about that. Completely forgot about that.